From 02f837f5d57fba9665df1c28f56a82e3f0eeeda2 Mon Sep 17 00:00:00 2001
From: rigaz29 <ryanbagas27@gmail.com>
Date: Mon, 27 Jul 2026 00:13:41 +0000
Subject: [PATCH] Adapt KSU static hooks for ReSukiSU

Titik hook bawaan Mi-Thorium ditulis untuk KernelSU generasi lama.
Empat penyesuaian agar cocok dengan ReSukiSU (Manual Hook):

1. fs/read_write.c: hapus blok hook vfs_read. ReSukiSU menandai
   ksu_vfs_read_hook sebagai incompatible dan menolak build; jalur read
   ditangani CONFIG_KSU_MANUAL_HOOK_AUTO_INITRC_HOOK lewat LSM.

2. fs/stat.c: tambah ksu_handle_newfstat_ret dan ksu_handle_fstat64_ret
   pada SYSCALL_DEFINE2(newfstat) dan SYSCALL_DEFINE2(fstat64).

3. kernel/reboot.c: tambah ksu_handle_sys_reboot pada
   SYSCALL_DEFINE4(reboot).

4. fs/exec.c: buang toggle ksu_execveat_hook dan cabang
   ksu_handle_execveat_sucompat -- keduanya menyebabkan undefined symbol.
---
 fs/exec.c       | 10 +++-------
 fs/read_write.c | 11 -----------
 fs/stat.c       | 12 ++++++++++++
 kernel/reboot.c |  8 ++++++++
 4 files changed, 23 insertions(+), 18 deletions(-)

diff --git a/fs/exec.c b/fs/exec.c
index 143098d39949..2218c20ae951 100644
--- a/fs/exec.c
+++ b/fs/exec.c
@@ -1896,11 +1896,10 @@ static int __do_execve_file(int fd, struct filename *filename,
 }
 
 #if IS_ENABLED(CONFIG_KSU_STATIC_HOOKS)
-extern bool ksu_execveat_hook __read_mostly;
+/* ReSukiSU tidak mengekspor toggle runtime, dan varian _sucompat hanya
+ * dikompilasi bila jalur SuSFS aktif. Panggil tanpa syarat. */
 extern int ksu_handle_execveat(int *fd, struct filename **filename_ptr, void *argv,
 			void *envp, int *flags);
-extern int ksu_handle_execveat_sucompat(int *fd, struct filename **filename_ptr,
-				 void *argv, void *envp, int *flags);
 #endif
 
 static int do_execveat_common(int fd, struct filename *filename,
@@ -1909,10 +1908,7 @@ static int do_execveat_common(int fd, struct filename *filename,
 			      int flags)
 {
 #if IS_ENABLED(CONFIG_KSU_STATIC_HOOKS)
-	if (unlikely(ksu_execveat_hook))
-		ksu_handle_execveat(&fd, &filename, &argv, &envp, &flags);
-	else
-		ksu_handle_execveat_sucompat(&fd, &filename, &argv, &envp, &flags);
+	ksu_handle_execveat(&fd, &filename, &argv, &envp, &flags);
 #endif
 	return __do_execve_file(fd, filename, argv, envp, flags, NULL);
 }
diff --git a/fs/read_write.c b/fs/read_write.c
index 27d438f49ff7..ec2d20ddd766 100644
--- a/fs/read_write.c
+++ b/fs/read_write.c
@@ -434,21 +434,10 @@ ssize_t kernel_read(struct file *file, void *buf, size_t count, loff_t *pos)
 }
 EXPORT_SYMBOL(kernel_read);
 
-#if IS_ENABLED(CONFIG_KSU_STATIC_HOOKS)
-extern bool ksu_vfs_read_hook __read_mostly;
-extern int ksu_handle_vfs_read(struct file **file_ptr, char __user **buf_ptr,
-			size_t *count_ptr, loff_t **pos);
-#endif
-
 ssize_t vfs_read(struct file *file, char __user *buf, size_t count, loff_t *pos)
 {
 	ssize_t ret;
 
-#if IS_ENABLED(CONFIG_KSU_STATIC_HOOKS)
-	if (unlikely(ksu_vfs_read_hook))
-		ksu_handle_vfs_read(&file, &buf, &count, &pos);
-#endif
-
 	if (!(file->f_mode & FMODE_READ))
 		return -EBADF;
 	if (!(file->f_mode & FMODE_CAN_READ))
diff --git a/fs/stat.c b/fs/stat.c
index 9f92d22b0a06..a143776948b6 100644
--- a/fs/stat.c
+++ b/fs/stat.c
@@ -150,6 +150,12 @@ EXPORT_SYMBOL(vfs_statx_fd);
 
 #if IS_ENABLED(CONFIG_KSU_STATIC_HOOKS)
 extern int ksu_handle_stat(int *dfd, const char __user **filename_user, int *flags);
+#ifdef CONFIG_KSU_MANUAL_HOOK
+extern void ksu_handle_newfstat_ret(unsigned int *fd, struct stat __user **statbuf_ptr);
+#if defined(__ARCH_WANT_STAT64) || defined(__ARCH_WANT_COMPAT_STAT64)
+extern void ksu_handle_fstat64_ret(unsigned long *fd, struct stat64 __user **statbuf_ptr);
+#endif
+#endif
 #endif
 
 /**
@@ -383,6 +389,9 @@ SYSCALL_DEFINE2(newfstat, unsigned int, fd, struct stat __user *, statbuf)
 	if (!error)
 		error = cp_new_stat(&stat, statbuf);
 
+#ifdef CONFIG_KSU_MANUAL_HOOK
+	ksu_handle_newfstat_ret(&fd, &statbuf);
+#endif
 	return error;
 }
 
@@ -509,6 +518,9 @@ SYSCALL_DEFINE2(fstat64, unsigned long, fd, struct stat64 __user *, statbuf)
 	if (!error)
 		error = cp_new_stat64(&stat, statbuf);
 
+#ifdef CONFIG_KSU_MANUAL_HOOK
+	ksu_handle_fstat64_ret(&fd, &statbuf);
+#endif
 	return error;
 }
 
diff --git a/kernel/reboot.c b/kernel/reboot.c
index 3a87fee47d3a..e22036d9d46e 100644
--- a/kernel/reboot.c
+++ b/kernel/reboot.c
@@ -305,6 +305,10 @@ DEFINE_MUTEX(system_transition_mutex);
  *
  * reboot doesn't sync: do that yourself before calling this.
  */
+#ifdef CONFIG_KSU_MANUAL_HOOK
+extern int ksu_handle_sys_reboot(int magic1, int magic2, unsigned int cmd, void __user **arg);
+#endif
+
 SYSCALL_DEFINE4(reboot, int, magic1, int, magic2, unsigned int, cmd,
 		void __user *, arg)
 {
@@ -312,6 +316,10 @@ SYSCALL_DEFINE4(reboot, int, magic1, int, magic2, unsigned int, cmd,
 	char buffer[256];
 	int ret = 0;
 
+#ifdef CONFIG_KSU_MANUAL_HOOK
+	ksu_handle_sys_reboot(magic1, magic2, cmd, &arg);
+#endif
+
 	/* We only trust the superuser with rebooting the system. */
 	if (!ns_capable(pid_ns->user_ns, CAP_SYS_BOOT))
 		return -EPERM;
-- 
2.43.0

